### PR TITLE
feat: WIP fast automata decision procedure for bitvectors

### PR DESCRIPTION
**Pull request by tobiasgrosser**
This is still work-in-progress, but the changes are contained to the `Experimental` folder and does not break other things.
Therefore, we're merging it to prevent it from bitrotting.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -34,7 +34,7 @@ jobs:
       - name: Compile Decide Experiment 🧐
         run: |
           lake -R exe cache get # download cache of mathlib docs.
-          lake -R build SSA.Experimental.Bits.Decide
+          lake -R build SSA.Experimental.Bits.Decide SSA.Experoimental.Bits.Fast.Decide
```
Comment (tobiasgrosser) [inline .github/workflows/ci.yml]:
```suggestion
          lake -R build SSA.Experimental.Bits.Decide SSA.Experimental.Bits.Fast.Decide
```

**Comment by tobiasgrosser** [comment]
@alexkeizer, I added some sorries to make this build clean. I suggest we merge this to keep this code up-to-date with lean updates.

**Comment by alexkeizer** [comment]
Agreed. It seems I've somehow broken the build again, I'll have a look at fixing that and then I'll hit merge

**Comment by tobiasgrosser** [comment]
Merge to ensure that we do not regress further.

**Comment by tobiasgrosser** [comment]
@alexkeizer, this seems to be still quite broken. Maybe it's worth looking at this tomorrow.

File: SSA/Experimental/Bits/Propagate.lean [hunk 2]
```diff
@@ -187,17 +188,17 @@ lemma id_eq_propagate (x : ℕ → Bool) :
...
       (λ f (_ : Empty → Bool) => (λ _ => false, f ())) Empty.elim := by
   ext n
   match n with
   | 0 => rfl
   | 1 => rfl
-  | n+2 => simp [oneSeq, propagate_succ]
+  | n+2 => simp [propagate_succ]
 
 lemma and_eq_propagate (x y : ℕ → Bool) :
     andSeq x y = propagate Empty.elim
```
Comment (tobiasgrosser) [inline SSA/Experimental/Bits/Propagate.lean]: @alexkeizer, it seems `andSeq` and friends must be updated to bitstreams.

**Comment by tobiasgrosser** [comment]
I merged the original fast_decide support, such that the diff here is now a lot smaller.

**Comment by AtticusKuhn** [comment]
What is lsSeq?

**Comment by AtticusKuhn** [comment]
Why is this PR on the merge queue? There are still `sorry`s in some of the files.

**Comment by tobiasgrosser** [comment]
Unsure. I may have misclicked.

File: SSA/Experimental/Bits/Fast/BitStream.lean [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+
+-- TODO: upstream the following section
+section UpStream
+
+namespace Int
+
+theorem bmod_eq_of_ge_and_le (z : Int) (m : Nat)
+    (lower_bound : m/2 ≤ z) (upper_bound : z < m/2) :
+    z.bmod m = z := by
+  omega
```
Comment (bollu) [inline SSA/Experimental/Bits/Fast/BitStream.lean]:
```suggestion
theorem bmod_eq_of_ge_and_le (z : Int) (m : Nat)
    (hlower_bound : m/2 ≤ z) (hupper_bound : z < m/2) :
    z.bmod m = z := by
  omega
```

File: SSA/Experimental/Bits/Fast/BitStream.lean [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+  cases decide (0 < w) <;> rfl
+
+end BitVec
+end UpStream
+
+
+
+
+
+
```
Comment (bollu) [inline SSA/Experimental/Bits/Fast/BitStream.lean:110]:
```suggestion

```

Kill dead whitespace

**Comment by bollu** [comment]
Go for it. I didn't read it carefully, I'd suggest just merging it and moving on.

**Pull request merged**